src/garage/torch/algos/td3.py

```diff
@@ -0,0 +1,286 @@
+"""TD3 model in Pytorch."""
+import copy
+
+from dowel import logger, tabular
+import numpy as np
+import torch
+
+from garage import _Default, log_performance, make_optimizer
+from garage.np import obtain_evaluation_samples
+from garage.np.algos import RLAlgorithm
+from garage.torch import dict_np_to_torch, torch_to_np
+
+
+class TD3(RLAlgorithm):
+    """Implementation of TD3.
+
+    Based on https://arxiv.org/pdf/1802.09477.pdf.
+
+    Args:
+        env_spec (garage.envs.EnvSpec): Environment.
+        policy (garage.torch.policies.Policy): Policy (actor network).
+        qf (garage.torch.q_functions.QFunction): Q-function (critic network).
+        qf2 (garage.torch.q_functions.QFunction): Q function (critic network).
+        replay_buffer (garage.replay_buffer.ReplayBuffer): Replay buffer.
+        target_update_tau (float): Interpolation parameter for doing the
+            soft target update.
+        discount (float): Discount factor (gamma) for the cumulative return.
+        reward_scaling (float): Reward scaling.
+        update_actor_interval (int): Policy (Actor network) update interval.
+        max_action (float): Maximum action magnitude.
+        buffer_batch_size (int): Size of replay buffer.
+        min_buffer_size (int): The minimum buffer size for replay buffer.
+        policy_noise (float): Policy (actor) noise.
+        policy_noise_clip (float): Noise clip.
+        clip_return (float): Clip return to be in [-clip_return,
+            clip_return].
+        policy_lr (float): Learning rate for training policy network.
+        qf_lr (float): Learning rate for training Q network.
+        policy_optimizer (Union[type, tuple[type, dict]]): Type of optimizer
+            for training policy network. This can be an optimizer type such as
+            `torch.optim.Adam` or a tuple of type and dictionary, where
+            dictionary contains arguments to initialize the optimizer
+            e.g. `(torch.optim.Adam, {'lr' : 1e-3})`.
+        qf_optimizer (Union[type, tuple[type, dict]]): Type of optimizer
+            for training Q-value network. This can be an optimizer type such
+            as `torch.optim.Adam` or a tuple of type and dictionary, where
+            dictionary contains arguments to initialize the optimizer
+            e.g. `(torch.optim.Adam, {'lr' : 1e-3})`.
+        steps_per_epoch (int): Number of train_once calls per epoch.
+        n_train_steps (int): Number of training steps.
```

Comment:
```suggestion
        grad_steps_per_env_step (int): Number of gradient steps taken per environment step sampled.
```

Please update the code to reflect this!